src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageLoaderMatcherPolicy.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    internal class PageLoaderMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        private readonly IPageLoader _loader;
+
+        public PageLoaderMatcherPolicy(IPageLoader loader)
+        {
+            if (loader == null)
+            {
+                throw new ArgumentNullException(nameof(loader));
+            }
+
+            _loader = loader;
+        }
+
+        public override int Order => int.MinValue + 100;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+            
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var page = endpoints[i].Metadata.GetMetadata<PageActionDescriptor>();
+                if (page != null)
+                {
+                    // Found a page
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+            
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                ref var candidate = ref candidates[i];
+                var endpoint = (RouteEndpoint)candidate.Endpoint;
+
+                var page = endpoint.Metadata.GetMetadata<PageActionDescriptor>();
+                if (page != null)
+                {
+                    var compiled = _loader.Load(page);
```

Comment:
Interesting. We could use this to fix - https://github.com/aspnet/AspNetCore/pull/6848 - specifically for endpoint routing

Reply:
A bit of a stretch because we don't need to do anything, but this might let us not use `IActionDescriptorChangeProvider` with endpoint routing. Essentially `IPageLoader.Load` will reflect the cache coherent version of the page. We could use this to replace the candidate rather than rely on `IADChangeProvider`

Reply:
I don't super-get your comment. but yes. You can do the async

Reply:
> A bit of a stretch because we don't need to do anything, but this might let us not use `IActionDescriptorChangeProvider` with endpoint routing.

That lets us add and remove as well as change though.